File: lib/models/item/story.dart
import 'package:hacki/config/constants.dart';
import 'package:hacki/models/item/item.dart';

class Story extends Item {
  const Story({
    required super.descendants,
    required super.id,
    required super.score,
    required super.time,
    required super.by,
    required super.title,
    required super.type,
    required super.url,
    required super.text,
    required super.kids,
    required super.parts,
    required super.hidden,
  }) : super(
          dead: false,
          deleted: false,
          parent: 0,
        );

  Story.empty() : super.empty();

  Story.placeholder()
      : super(
          id: 0,
          score: 0,
          descendants: 0,
          time: 1171872000,
          by: 'Y Combinator',
          title: 'Hacker News Guidelines',
          url: Constants.guidelineLink,
          kids: <int>[],
          dead: false,
          parts: <int>[],
          deleted: false,
          parent: 0,
          text: '',
          type: '',
          hidden: false,
        );

  Story.fromJson(super.json) : super.fromJson();

  Story copyWith({bool? hidden}) {
    return Story(
      descendants: descendants,
      id: id,
      score: score,
      time: time,
      by: by,
      title: title,
      type: type,
      url: url,
      text: text,
      kids: kids,
      parts: parts,
      hidden: hidden ?? this.hidden,
    );
  }

  String get metadata =>
      '''$score point${score > 1 ? 's' : ''}${by.isNotEmpty ? ' $by ' : ' '}$timeAgo | $descendants comment${descendants > 1 ? 's' : ''}''';

  String get screenReaderLabel =>
      '''$title, at $readableUrl, by $by $timeAgo. This story has $score point${score > 1 ? 's' : ''} and $descendants comment${descendants > 1 ? 's' : ''}''';

  String get simpleMetadata =>
      '''$score point${score > 1 ? 's' : ''} $descendants comment${descendants > 1 ? 's' : ''} $timeAgo''';

  String get readableUrl {
    final Uri url = Uri.parse(this.url);
    final String authority = url.authority.replaceFirst('www.', '');
    return authority;
  }

  @override
  String toString() => 'Story $id';
}

File: ios/StoryWidget/Models/Story.swift
public extension Story {
    var shortMetadata: String {
        if isJob {
            return "\(timeAgo)"
        } else {
            return "\(score.orZero) | \(descendants.orZero) | \(timeAgo)"
        }
    }
}

public struct Story: Item {
    public let id: Int
    public let parent: Int?
    public let title: String?
    public let text: String?
    public let url: String?
    public let type: String?
    public let by: String?
    public let score: Int?
    public let descendants: Int?
    public let time: Int
    public let kids: [Int]?
    public var metadata: String {
        if isJob {
            return "\(timeAgo) by \(by.orEmpty)"
        } else {
            return "\(score.orZero) pt\(score.orZero > 1 ? "s":"") | \(descendants.orZero) cmt\(descendants.orZero > 1 ? "s":"") | \(timeAgo) by \(by.orEmpty)"
        }
    }

    public init(id: Int, 
                parent: Int? = nil,
                title: String?,
                text: String?,
                url: String?,
                type: String?,
                by: String?,
                score: Int?,
                descendants: Int?,
                time: Int,
                kids: [Int]? = [Int]()) {
        self.id = id
        self.parent = parent
        self.title = title
        self.text = text
        self.url = url
        self.type = type
        self.score = score
        self.by = by
        self.descendants = descendants
        self.time = time
        self.kids = kids
    }

    // Empty initializer
    public init() {
        self.init(
            id: 0,
            parent: 0,
            title: "",
            text: "",
            url: "",
            type: "story",
            by: "",
            score: 0,
            descendants: 0,
            time: 0
        )
    }

    func copyWith(text: String?) -> Story {
        .init(
            id: id,
            parent: parent,
            title: title,
            text: text,
            url: url,
            type: type,
            by: by,
            score: score,
            descendants: descendants,
            time: time,
            kids: kids
        )
    }
    
    public static let errorPlaceholder: Story = .init(
        id: 0,
        title: "Something went wrong...",
        text: nil,
        url: "retrying...",
        type: "story",
        by: nil,
        score: nil,
        descendants: nil,
        time: 0
    )
}
